src/cargo/core/resolver/resolve.rs

```diff
@@ -41,6 +44,21 @@ impl Resolve {
         unused_patches: Vec<PackageId>,
     ) -> Resolve {
         let reverse_replacements = replacements.iter().map(|(&p, &r)| (r, p)).collect();
+        let public_dependencies = graph.iter().map(|p| {
+            let public_deps = graph.edges(p).flat_map(|(dep_package, deps)| {
+                let id_opt: Option<PackageId> = deps.iter().find(|d| d.kind() == Kind::Normal).and_then(|d| {
```

Comment:
Just to be sure I understand, this calculates the direct (not transitive) public dependencies for each package. Yes?
Why the the `d.kind() == Kind::Normal`?

Reply:
The other kinds are `Development` and `Build`, neither of which can ever be exposed as transitive dependencies.

Reply:
That makes sense. Should we have an error if you try a Cargo.toml that has a `Development` or `Build` dep with public set to true? Then add asserts in the `set_public` to make sure we never let it slip in?